src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceCollectionDescriptorExtensionsTests.cs

```diff
@@ -2,15 +2,54 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+using System.Linq;
 using Microsoft.Extensions.DependencyInjection.Extensions;
 using Microsoft.Extensions.DependencyInjection.Specification.Fakes;
+using Microsoft.Extensions.DependencyInjection.Tests;
 using Xunit;
-using Xunit.Abstractions;
 
 namespace Microsoft.Extensions.DependencyInjection
 {
+    public class ABC { }
+
     public class ServiceCollectionDescriptorExtensionsTest
     {
+        [Fact]
+        internal void GetService_FactoryCallSite_Transient_DoesNotFail()
```

Comment:
Would this test catch the bug you are fixing?

Reply:
yes it does

Reply:
Because the transient service is being registered as a factory callsite, then ImplementationType will be null and will cause ArgumentNullException and cause Debug.Fail and that log to get fired.

Reply:
FactoryCallSite has a null implementation type so it throws argument null here.

Reply:
That's a pretty indirect way to test it.

We can test it more directly by registering 

```C#
IService -> Service
Service -> Service
```

And ensuring that both accessors return distinct values.

What I don't understand is don't we already have a test that uses a transient factory using a dynamic service provider? Or is the problem that we don't actually get to compile it in those tests? 

Should we add a mode that uses the dynamic engine but forces the compilation eagerly and run all tests through it?

Reply:
OK I'll check where these similar tests are. I updated the PR description to describe the two scenarios being fixed here.

Reply:
I added a test that does the two above registrations with transient lifetime and using factory callsite, kicks off background thread and after the fix Debug.Fail, won't hit anymore.